src/ripple/app/tx/impl/Payment.cpp

```diff
@@ -346,12 +347,30 @@ Payment::doApply ()
         view().update (sleDst);
     }
 
-    TER terResult;
+    TER terResult {tesSUCCESS};
+
+    // Determine whether the destination requires deposit authorization.
+    bool const reqDepositAuth = [] (ReadView& view, std::uint32_t dstFlags)
+        {
+            if (dstFlags & lsfDepositAuth &&
+                view.rules().enabled(featureDepositAuth))
+                    return true;
+
+            return false;
+        } (view(), sleDst->getFlags());
 
     bool const bRipple = paths || sendMax || !saDstAmount.native ();
     // XXX Should sendMax be sufficient to imply ripple?
 
-    if (bRipple)
+    if (bRipple && reqDepositAuth)
+    {
+        // If
+        //  1. payment is not direct XRP to XRP and
+        //  2. the destination has lsfDepositAuth set,
+        // then the payment is not authorized.
+        terResult = tecNO_PERMISSION;
```

Comment:
Instead of setting `terResult` would could return here (and in other places in this function). I think doing to simplifies the code. I took a cut at a refactor here: https://github.com/seelabs/rippled/commit/d6efff900be2c420fdeef5b527fa1114b6bb2088